core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1624,6 +1630,47 @@ private RexNode simplifyOrs(List<RexNode> terms, RexUnknownAs unknownAs) {
             continue;
           }
         }
+        break;
+      case NOT_EQUALS:
+        final Comparison notEqualsComparison =
+            Comparison.of(term, node -> RexUtil.isReferenceOrAccess(node, false));
+        if (notEqualsComparison != null) {
+          // We are dealing with a X<>A term, check if we saw before another NOT_EQUALS involving X
+          final RexNode prevNotEquals = notEqualsComparisonMap.get(notEqualsComparison.ref);
+          if (prevNotEquals == null) {
+            // This is the first NOT_EQUALS involving X, put it in the map
+            notEqualsComparisonMap.put(notEqualsComparison.ref, term);
+          } else {
+            // There is already in the map another NOT_EQUALS involving X:
+            //   - if it is already an IS_NOT_NULL: it was already simplified, ignore this term
+            //   - if it is not an IS_NOT_NULL (i.e. it is a NOT_EQUALS): check comparison values
+            if (prevNotEquals.getKind() != SqlKind.IS_NOT_NULL) {
+              final Comparable comparable1 = notEqualsComparison.literal.getValue();
+              //noinspection ConstantConditions
+              final Comparable comparable2 = Comparison.of(
+                  prevNotEquals, node -> RexUtil.isReferenceOrAccess(node, false))
+                  .literal.getValue();
+              //noinspection unchecked
+              if (comparable1.compareTo(comparable2) != 0) {
```

Comment:
when `comparable1.compareTo(comparable2) == 0`, does it make sense to remove the identical one ?

Reply:
I think this case shall never occurred, because such simplification should have been done earlier by `simplifyComparison`. 
In any case, if we face such a scenario, the "second" identical not equals comparison will be disregarded and effectively removed from the final list of terms, by the following code below that if block:
```
terms.remove(i);
--i;
continue;
```

Reply:
THX